Randomize packaging upgrade test

jasontedor opened the pull request:
This commit adds randomization for the packaging upgrade test. In
particular, we extract a list of the released version of Elasticsearch
from Maven Central and randomize the selection of the version to upgrade
from. The randomization is repeatable, and supports the tests.seed
property. Specific versions can be tested by setting the property
tests.packaging.upgrade.from.versions.

jasontedor committed:
Randomize packaging upgrade test

This commit adds randomization for the packaging upgrade test. In
particular, we extract a list of the released version of Elasticsearch
from Maven Central and randomize the selection of the version to upgrade
from. The randomization is repeatable, and supports the tests.seed
property. Specific versions can be tested by setting the property
tests.packaging.upgrade.from.versions.

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -56,9 +56,32 @@ for (String box : getProperties().get('vagrant.boxes', 'sample').split(',')) {
+  upgradeFromVersions = maybeUpdradeFromVersions
+} else {
+  new URL('http://repo1.maven.org/maven2/org/elasticsearch/elasticsearch/maven-metadata.xml').openStream().withStream { s ->
+    def xml = new XmlParser().parse(s)
```
We often try to avoid `def` in these where possible.

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -56,9 +56,32 @@ for (String box : getProperties().get('vagrant.boxes', 'sample').split(',')) {
+if (maybeUpdradeFromVersions != null) {
+  upgradeFromVersions = maybeUpdradeFromVersions
+} else {
+  new URL('http://repo1.maven.org/maven2/org/elasticsearch/elasticsearch/maven-metadata.xml').openStream().withStream { s ->
```
I believe this happens as soon as you run gradle and it starts initializing the projects. I'm not sure we should do network IO then. Either we should keep the list locally and implement a gradle task to fetch it or, well, we should do something else smarter than what I can think of right now.

nik9000 replied:
If you made all of this into a method that you execute lazily, say by interpolating it in a string, then maybe this kind of thing is ok.

rjernst replied:
Yes as this is written now it will do network IO every time gradle is invoked, regardless of what task is specified. We should start with a list of versions here.

jasontedor committed:
Add static list of upgrade from versions

This commit adds a resource containting a static list of versions to
randomly upgrade from in the packaging tests. This resource is added to
avoid doing network I/O during gradle initialization. Instead, this list
of versions is validated against a reliable list of versions only when a
packaging test task is executed. Also provided is a task to
automatically update this resource should it go out of date.

jasontedor commented:
Thanks for the reviews @nik9000 and @rjernst. I added a static resource containing a list of versions to randomly upgrade from. This list of versions will only be validated when an actual packaging test task is executed. I also provided a task to automatically update this list of versions should it go out of date (similar to `gradle updateSHAs`). Let me know what you think?

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -92,6 +71,24 @@ repositories {
+  if (maybeUpdradeFromVersions != null) {
+    upgradeFromVersions = maybeUpdradeFromVersions.split(",")
+  } else {
+    upgradeFromVersions = new File(project.projectDir, 'src/main/resources/versions')
```
nit: *Technically* this belongs in `src/build/resources/version` or some place like that. Just not *main*.

nik9000 replied:
I'm fairly sure the dependencies closure is executed immediately and we try not to do IO there. But @rjernst will know what is the most right thing to do here. It could be as simple as making a method for this and calling it from within the string on line 104 and 105 with `${ -> upgradeFromVersion() }`

rjernst replied:
Nik is right, but I think his idea might work to delay until we need to resolve.

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -129,8 +115,34 @@ task stop {
   description 'Stop any tasks from tests that still may be running'
 }
 
+def getVersions() {
```
s/def/String<String>/. This is just how we've been writing groovy. Mostly.

jasontedor committed:
Add type for helper method in vagrant/build.gradle

This commit adds strong typing for the getVersions helper method in
vagrant/build.gradle.

jasontedor committed:
Move Vagrant resource to src/build/resources

This commit moves the build resource file versions used in the Vagrant
tests to src/build/resources from src/main/resources.

nik9000 commented:
OK! I had a look. I resolved a few issues with the gradle but wasn't able to fix the IO-at-init-time issue. It turns out that gradle doesn't allow dependencies to be resolved asynchronously. So we'd have to download the dependency from maven central and handle the caching ourselves. Which is icky!

This is where I got: https://github.com/jasontedor/elasticsearch/compare/randomize-upgrade-test...nik9000:pr/19033?expand=1

jasontedor committed:
Fix return type in helper in vagrant/build.gradle

This commit fixes a return type mistake introduced in commit
37bc3c4de360ec4b7a932ec0b35866e1b1a1837e.

jasontedor committed:
Restore lost versions file in upgrade test

This commit restores a lost versions file that was lost due to
.gitignore in commit 4ba2db44d6535c5332829e046e347bbb9a674cad.

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -95,8 +115,33 @@ task stop {
+}
+
+task verifyPackagingTestUpgradeFromVersions {
+  String maybeUpdateFromVersions = System.getProperty("tests.packaging.upgrade.from.versions", null)
```
The way this is written now it happens at task construction time, not task execution time. So this is done every time gradle is started. If you wrap it in `doLast` then it'll wait. The same is true of the above task.

jasontedor replied:
I pushed 32a3d0d623600aa18d9766dffd9ae78b2d6b94ff.

nik9000 commented on qa/vagrant/build.gradle:
```diff
@@ -69,6 +71,24 @@ repositories {
+    upgradeFromVersions = new File(project.projectDir, 'versions')
+  }
+
+  upgradeFromVersion = upgradeFromVersions[new Random(seed).nextInt(upgradeFromVersions.length)]
```
Putting these in the dependencies closure doesn't really do anything. I'd probably just move them back up to the top.

I don't think we have any chance of avoiding the IO on startup given how gradle handles depdencies. I think it is worth it so we can have this randomization though.

jasontedor replied:
I pushed 3111314bc21be6022a945ba78928859bced5e6cf.

jasontedor committed:
Initialize upgrade version outside of dependencies

jasontedor committed:
Do not get versions until task execution

jasontedor commented:
@nik9000 I've responded to your feedback.

nik9000 commented:
LGTM. I'm sad to see us having to load the file on script initialization but I don't see a way around it and I want this.
